Refactoring WCPayCard into new onboarding package

joelclimbsthings opened the pull request:
Fixes woocommerce/woocommerce-admin#6868

Creates a new package, `@woocommerce/tasks` to house the refactored WCPayCard components. These will be consumed by the WCPay repo, as well as used on the payments tasks page as it was previously. They've been refactored to aid in flexibility for how they're consumed. 

I've also moved the slotfill components into this package, so we'll likely need to publish this for the `remote-payments-tester` to work properly. The alternative is using `npm link` in the short-term.

Also included are other revisions and tweaks to the logic around `wcpay` to ensure that its working as expected.

_Other Note:_ There was a little discussion on how "broad" to make this new package, with talk of having a more general one around `onboarding` in total. After Sam's [recent p2 post](https://teamisotope.wordpress.com/2021/05/13/woocommerce-components-is-too-big/) I thought something a bit smaller and more granular might better suit the current trend here.

### Open questions
- I initially had the new tasks package added to [the DEWP config](https://github.com/woocommerce/woocommerce-admin/blob/main/packages/dependency-extraction-webpack-plugin/assets/packages.js), but I eventually figured out is was causing an issue and the react app wasn't loading at all. I can revisit, but wasn't sure if this was necessary for this package?

### Screenshots

![image](https://user-images.githubusercontent.com/444632/119566468-08f95480-bd60-11eb-84c3-82959ae12320.png)


### Detailed test instructions:

-  Enable remote payments and ensure your opted into marketing
-  Clone the `try/use-tasks-package` branch of [my fork of the remote payment tester](https://github.com/joelclimbsthings/woocommerce-admin-remote-tester)
- Set your store to a US based location
- Use `yarn link` to link your remote payment tester to the new `@woocommerce/tasks` package
- Ensure that the WCPay Card appears as pictured when you navigate to the payment task
- Click on "Get Started," and verify that the slot-fill works. Note that no fields have been setup, so it will just display the "Proceed" button for now.

joelclimbsthings committed:
Refactoring WCPayCard into tasks package, along with payment task slotfills

joelclimbsthings committed:
Adding changelog

joshuatf commented on client/task-list/tasks/payments/RemotePayments/index.js:
```diff
@@ -38,7 +39,7 @@ export const RemotePayments = ( { query } ) => {
 		const method = methods.find(
 			( m ) =>
-				( m.key === 'wcpay' && m.visible ) ||
+				( m.key === 'woocommerce_payments' && m.visible ) ||
```
Heads up that these have been moved into a constant array `RECOMMENDED_GATEWAY_KEYS`.

joelclimbsthings replied:
Cool, no worries here. Just updating references. Note that the logic/ui for WCPay is going to be different than other recommended gateways for all except the final stage (enabled + configured) so we may need more iteration here.

joshuatf reviewed (commented):
Great work here, Joel!  Left a couple questions just to get some conversation going, but otherwise most comments were about conflicts.  I also suspect we'll a few more with https://github.com/woocommerce/woocommerce-admin/pull/7073 but we can get this merged first and I'll take care of the conflicts there.

joshuatf commented on client/task-list/tasks/payments/components/WCPayMethodCard.js:
```diff
@@ -0,0 +1,99 @@
+
+	return (
+		<WCPayCard>
+			<CardHeader isEnabled={ isEnabled } />
```
Same comments on this re: other components in this file.

samueljseay replied:
Naive jump in on this when I see props like `headerComponent`. can we please stop using a pattern of passing components as props. Use children, its easier to reuse/compose/refactor

joelclimbsthings replied:
> Not sure if we have a team opinion on this, but the extra component CardHeader seems a little redundant and I end up scrolling up and down quite a bit to piece things together.

I go back and forth on this. I actually find it easier to glean the overall structure when composed in this way, although it's probably pretty unnecessary in this case since the components are small. Consider it refactored.

joelclimbsthings commented on packages/tasks/README.md:
```diff
@@ -0,0 +1,11 @@
+# Tasks
```
Totally fair, and I mentioned this briefly in the _Other Note_ portion of the description above. @psealock actually mentioned the potential of an `onboarding` package, but it seemed that there may be benefits to erring on the side of smaller, more granular package given Sam's [recent post](https://teamisotope.wordpress.com/2021/05/13/woocommerce-components-is-too-big/). I'm not all that opinionated either way.

samueljseay replied:
@joelclimbsthings if it helps at all @louwie17 recently worked on enabling tree shaking for `@woocommerce/components` so these could go there if you prefer.

joelclimbsthings replied:
Thanks @samueljseay , I think this is debatable but I'd just as soon keep it here for now.

psealock replied:
+1 for `onboarding`

joelclimbsthings replied:
Ohhhh alright then. Package name changed to `onboarding`.

joelclimbsthings committed:
Moving pill option out of the RecommendedRibbon component, and minor refactor

samueljseay commented on packages/tasks/src/components/RecommendedRibbon.js:
```diff
@@ -2,20 +2,19 @@
 
-	return (
-		<div className={ classes }>
+	return isPill ? (
```
Any reason why we're not starting out with `ts` on new components? It would really help if we could start with typescript rather than relying on refactoring to get to it later.

joelclimbsthings replied:
> Maybe we can simplify this to:

Thanks @joshuatf , I actually didn't like having a `<Pill />` returned from a component called `<RecommendationRibbon>`, so I ended up factoring this out of it. 

> Any reason why we're not starting out with ts on new components? 

@samueljseay Really just due to the fact that it would take me twice as long at this stage, and we're already having to bring in members from other teams to try to get this over the finish line this sprint. If we decide as a team to prioritize TS over speed in the short/medium term, then I'm all for it. Otherwise I'm just approaching it as time allows or my own stubbornness insists.

samueljseay replied:
@joelclimbsthings totally fair enough :) I appreciate you considering the pragmatic path forward 👍

joelclimbsthings commented on client/task-list/tasks/payments/components/WCPayMethodCard.js:
```diff
@@ -0,0 +1,99 @@
+const TosPrompt = () =>
+	interpolateComponents( {
+		mixedString: __(
+			'Upon clicking "Get started", you agree to the {{link}}Terms of service{{/link}}. Next we’ll ask you to share a few details about your business to create your account.',
```
I think that will result it being displayed in the `connect` step as well, in which case the "Upon clicking Get started" bit won't really make sense. This surfaces one unknown for me. Is something like this what we _want_ displayed during the connect step for WCPay, and what do we reference to figured this out?

joshuatf replied:
Sorry, you can ignore my comment.  It should just be the text in the connection step above the submit button, not on the card shown directly above the list.

joelclimbsthings commented on client/task-list/tasks/payments/components/WCPayMethodCard.js:
```diff
@@ -0,0 +1,99 @@
+const CardBody = () => (
+	<WCPayCardBody
+		description={ __(
+			'Try the new way to get paid. Securely accept credit and debit cards on your site. Manage transactions without leaving your WordPress dashboard. Only with WooCommerce Payments. ',
```
Good call, I'm assuming you're referring to a `description` property set on the configuration in WCCOM? I've made this change, and it's currently working with the dummy configuration from the Remote Payments Tester.

joelclimbsthings committed:
Fixing isEnabled status in WCPaymentMethodCard and a minor refactor

joelclimbsthings committed:
Using children for WCPayCardHeader

joelclimbsthings commented:
Thanks for the thoughtful reviews @joshuatf  + @samueljseay . I believe I've addressed all of the feedback, with one or two open questions yet. Let me know what you think.

joelclimbsthings committed:
Renaming tasks package to onboarding

joelclimbsthings commented:
Changed package name to `onboarding` to appease the masses! After rebasing on `main` there are a few issues that cropped up, but at this point I think it's best to save these fixes for other pending PRs to be merged, so I'll revisit during the `wcpay` integration on [this issue](https://github.com/woocommerce/woocommerce-admin/issues/6878).

joshuatf commented on packages/onboarding/src/components/RecommendedRibbon/RecommendedRibbon.js:
```diff
@@ -5,17 +5,13 @@ import { __ } from '@wordpress/i18n';
-		? 'woocommerce-task-payment__recommended-pill'
-		: 'woocommerce-task-payment__recommended-ribbon';
-
+export const RecommendedRibbon = ( { methodKey } ) => {
```
Nice refactor 👍

joshuatf commented on packages/onboarding/src/components/WCPayCard/WCPayCard.js:
```diff
@@ -0,0 +1,65 @@
+	tosComponent: TosComponent = null,
+} ) => (
+	<CardFooter>
+		<TosComponent />
```
Could we use `children` here instead of passing components as props?

joelclimbsthings replied:
Yup, good call.

joshuatf commented on client/task-list/tasks/payments/components/WCPayMethodCard.js:
```diff
@@ -0,0 +1,92 @@
+
+			<WCPayCardBody
+				description={ description }
+				linkOnClick={ () => {
```
Nitpicky, but I think this should be `onLinkClick`.

joshuatf reviewed (approved):
Well done, Joel!  Thanks for answering questions and providing feedback.

Left a few relatively minor comments, but pre-approving so we can unblock things.

I'm also curious if there are benefits from `WCPayCardFooter` which is essentially a simple wrapper around `CardFooter` if we use `children`.

joelclimbsthings committed:
Implementing feedback

joelclimbsthings commented:
Thanks @joshuatf ! 

> I'm also curious if there are benefits from WCPayCardFooter which is essentially a simple wrapper around CardFooter if we use children.

That's entirely true at this point. I'm inclined to leave the abstraction there for the moment, or at least until I do the integration into the WCPay repo and see if it requires any changes. I also think it's somewhat useful to be able to import all required components from the same package.

joelclimbsthings committed:
Updating readme
